File: packing_interface/src/algorithm_templates/python/root/bfdh.py
import packing_lib
from typing import List, Tuple


class Packing:
    def solve(self, bin_width: int, rectangles: List[Tuple[int, int, int]]) -> List[Tuple[float, float, int, int]]:
        items = packing_lib.sort_by_height(packing_lib.expand_items(rectangles))
        levels = []
        placements = []
        current_y = 0

        for item in items:
            width = item["width"]
            height = item["height"]

            if width > bin_width:
                raise ValueError(f"Rectangle width {width} exceeds bin width {bin_width}.")

            best_level = None
            best_remaining_width = None

            for level in levels:
                if level["used_width"] + width > bin_width:
                    continue

                remaining_width = bin_width - (level["used_width"] + width)
                if best_remaining_width is None or remaining_width < best_remaining_width:
                    best_level = level
                    best_remaining_width = remaining_width

            if best_level is not None:
                placements.append((best_level["used_width"], best_level["y"], width, height))
                best_level["used_width"] += width
                continue

            levels.append({"y": current_y, "height": height, "used_width": width})
            placements.append((0, current_y, width, height))
            current_y += height

        return packing_lib.output_from_placements(bin_width, placements)

File: packing_interface/src/algorithm_templates/cpp/root/bfdh.cpp
#include "packing_lib.h"
using namespace packing;

class Packing {
public:
    std::vector<std::tuple<double, double, int, int>> solve(
        int binWidth,
        const std::vector<std::tuple<int, int, int>>& rectangles
    ) {
        auto items = sort_by_height(expand_items(rectangles));

        struct Level {
            int y;
            int height;
            int usedWidth;
        };

        std::vector<Level> levels;
        std::vector<std::tuple<double, double, int, int>> placements;
        int currentY = 0;

        for (const auto& item : items) {
            int best_level = -1;
            int best_remaining_width = 0;

            for (int i = 0; i < static_cast<int>(levels.size()); ++i) {
                const auto& level = levels[i];
                if (level.usedWidth + item.width > binWidth) {
                    continue;
                }

                int remaining_width = binWidth - (level.usedWidth + item.width);
                if (best_level == -1 || remaining_width < best_remaining_width) {
                    best_level = i;
                    best_remaining_width = remaining_width;
                }
            }

            if (best_level != -1) {
                auto& level = levels[best_level];
                placements.push_back({
                    static_cast<double>(level.usedWidth),
                    static_cast<double>(level.y),
                    item.width,
                    item.height,
                });
                level.usedWidth += item.width;
                continue;
            }

            levels.push_back({currentY, item.height, item.width});
            placements.push_back({0.0, static_cast<double>(currentY), item.width, item.height});
            currentY += item.height;
        }

        return placements;
    }
};
